cache.go

```diff
@@ -0,0 +1,140 @@
+	// sqlair.Statement they built from and the sqlair.DB they are prepared on.
+	stmtDBCache map[uint64]map[uint64]*sql.Stmt
+
+	// dbStmtCache indicates when a sqlair.Statement has been prepared on a particular sqlair.DB.
+	dbStmtCache map[uint64]map[uint64]bool
+
+	// stmtIDCount and dbIDCount are monotonically increasing counters used to
+	// generate unique new cache IDs.
+	stmtIDCount uint64
+	dbIDCount   uint64
+
+	mutex sync.RWMutex
+}
+
+var once sync.Once
+var singleStmtCache *statementCache
+
+// newStatementCache returns the single instance of the statement cache.
+func newStatementCache() *statementCache {
+	once.Do(func() {
+		singleStmtCache = &statementCache{
+			stmtDBCache: map[uint64]map[uint64]*sql.Stmt{},
+			dbStmtCache: map[uint64]map[uint64]bool{},
+		}
+	})
+	return singleStmtCache
+}
+
+// newStatement returns a new sqlair.Statement and adds it to the cache. A
+// finalizer is set on the sqlair.Statement to remove its ID from the cache and
+// close all associated sql.Stmt objects.
+func (sc *statementCache) newStatement(te *expr.TypeBoundExpr) *Statement {
+	cacheID := atomic.AddUint64(&sc.stmtIDCount, 1)
+	sc.mutex.Lock()
+	sc.stmtDBCache[cacheID] = map[uint64]*sql.Stmt{}
+	sc.mutex.Unlock()
+	s := &Statement{te: te, cacheID: cacheID}
+	// This finalizer is run after the Statement is garbage collected.
+	runtime.SetFinalizer(s, sc.removeAndCloseStmtFunc)
+	return s
+}
+
+// newDB returns a new sqlair.DB and allocates the necessary resources in the
+// statementCache. A finalizer is set on the sqlair.DB to remove references to
+// it from the cache, close all sql.Stmt objects on it, and close the sql.DB
+func (sc *statementCache) newDB(sqldb *sql.DB) *DB {
+	cacheID := atomic.AddUint64(&sc.dbIDCount, 1)
+	sc.mutex.Lock()
+	sc.dbStmtCache[cacheID] = map[uint64]bool{}
+	sc.mutex.Unlock()
+	db := &DB{sqldb: sqldb, cacheID: cacheID}
+	// This finalizer is run after the DB is garbage collected.
+	runtime.SetFinalizer(db, sc.removeAndCloseDBFunc)
+	return db
+}
+
+// lookupStmt checks if a *sql.Stmt corresponding to s has been prepared on db
+// and stored in the cache.
+func (sc *statementCache) lookupStmt(db *DB, s *Statement) (*sql.Stmt, bool) {
+	// The Statement cache ID is only removed from stmtDBCache when the
+	// finalizer is run. The Statements cache ID must be in the stmtDBCache
+	// since we hold a reference to the Statement. It is therefore safe to
+	// access in it in the map without first checking it exists.
+	sc.mutex.RLock()
+	sqlstmt, ok := sc.stmtDBCache[s.cacheID][db.cacheID]
+	sc.mutex.RUnlock()
+	return sqlstmt, ok
+}
+
+// storeStmt stores a prepared *sql.Stmt in the cache.
+func (sc *statementCache) storeStmt(db *DB, s *Statement, sqlstmt *sql.Stmt) {
+	sc.mutex.Lock()
+	// Check if sqlstmt has been inserted by another process else already.
+	sqlstmtAlt, ok := sc.stmtDBCache[s.cacheID][db.cacheID]
+	if ok {
+		sqlstmt.Close()
+		sqlstmt = sqlstmtAlt
```

Comment:
How are you testing this? 

If you want `sqlstmtAlt` to point to `sqlstmt` then you either need to write to the pointer `*sqlstmt` or return the `sql.Stmt`.

Reply:
Ah good point, that's just closing the original statement and doing nothing else. I've refactored this function to actually prepare the statement as well to simplify this a bit.

This is not being tested at the moment, the difficult thing is, this case is only triggered by a race. In general we haven't been testing for race issues and relying on inspection, but this was still very much missed.

We could add in more infrastructure to the test driver to test this sort of thing, adding artificial waits and so on, but for now I'm not that keen to put in the work to add it unless we start coming across issues with this elsewhere.

Reply:
That's better. You're still doing work to then throw it away if another goroutine got there first. Although this is likely to be an edge case, I wonder if this is still the best approach, or if another strategy can be used?

Reply:
imo this is very much an edge case, it has to be the same statement being used separately in two different places, i.e. two instances of exactly the same SQL query happening at the same time. Plus this code is only going to be hit if its the first time they are being prepared, any subsequent times the lookup will find the sql.Stmt in the cache and this function won't even be called.

An obvious alternative would just be to lock the cache for the whole preparation as well but that has a far higher cost since its a global cache and it will block all queries whilst the cache is locked and it will happen fairly often since running a statement for the first time on the database is a relatively common action.